### DISCUSSION TITLE
How to receive data from a Bluetooth LE UART service

### DISCUSSION THREAD

**Discussion by E812** [shinyorg/shiny#853] [Q&A]
I am trying to connect to an Adafruit Bluefruit LE-UART Friend and read data from it in a Xamarin Forms project on Android.
I first did some tests with the Plugin.BluetoothLE and was able to connect and receive data with the physical devices (code provided below). But then I noticed that there is Shiny with all the nice benefits, and as I am still in the very beginning of my project, I would like to switch over to shiny.
I managed to connect to the device, but I cannot get the data receiving part working. Compared to the Plugin.BluetoothLE API there is so much more available, but partly the documentation is not yet available and the samples are still in todo state.

This is what I had in Plugin.BluetoothLE:

```
private readonly Guid UartService = new Guid("6E400001-B5A3-F393-E0A9-E50E24DCCA9E");
private readonly Guid RxdCharacteristic = new Guid("6E400003-B5A3-F393-E0A9-E50E24DCCA9E");
private readonly Guid TxdCharacteristic = new Guid("6E400002-B5A3-F393-E0A9-E50E24DCCA9E");

private async void UseBluetoothLowEnergy()
{
    var locationPermission = new Permissions.LocationWhenInUse();
    var status = await this.CheckAndRequestPermissionAsync(locationPermission);
    if (status != PermissionStatus.Granted)
    {
        var newStatus = await locationPermission.RequestAsync();
        if (newStatus != PermissionStatus.Granted)
        {
            // Show permission not granted notification.

            return;
        }
    }

    var scanConfig = new ScanConfig
    {
        ServiceUuids = new[] { this.UartService }.ToList(),
    };

    CrossBleAdapter.Current.Scan(scanConfig).Subscribe(scanResult =>
    {
        var device = scanResult.Device;
        
        var connectionConfig = new ConnectionConfig { AutoConnect = true };
        if (device.Status != ConnectionStatus.Connected)
        {
            device.Connect(connectionConfig);

            device.ConnectHook(this.UartService, this.RxdCharacteristic).Subscribe(characteristicGattResult =>
            {
                var data = characteristicGattResult.Data;
				
                // Do someting with the received data...
            }
            });
        }

    });
}

public async Task<PermissionStatus> CheckAndRequestPermissionAsync<T>(T permission)
    where T : BasePermission
{
    var status = await permission.CheckStatusAsync();
    if (status != PermissionStatus.Granted)
    {
        status = await permission.RequestAsync();
    }

    return status;
}

```

Can you provide a sample how to achive the same with Shiny?

**Comment by aritchie** [comment]
The samples for Bluetooth have enough done for you to look at.

**Comment by E812** [comment]
Hi Allen,
I had a look at the samples before I posted here. The BluetoothLE-Hosting is not what I am looking for. Then I had a look at the BluetoothLE-Managed, but it only contains the code for how to scan for peripherals and display them.
The last sample I looked at is the  BluetoothLE-Client. 
When I try to run that on my device nothing happens. However I looked into the code. In the ScanViewModel there is code how to scan for peripherals. However the created PeripheralItemViewModel does not contain any logic how to go on from here. There are only display properties, no commands. I also checked if there would be any places where the CharacteristicViewModels would be created but did not find any. Also there are no delegates that would be triggered by the 3 commands on the CharacteristicViewModel.

Did I miss something?

For completeness here is what I tried with Shiny (following the managed approach):

```
public class AboutViewModel : BaseViewModel, IDisposable
    {
        private readonly string UartService = "6E400001-B5A3-F393-E0A9-E50E24DCCA9E".ToLower();
        private readonly string RxdCharacteristic = "6E400003-B5A3-F393-E0A9-E50E24DCCA9E".ToLower();
        private readonly string TxdCharacteristic = "6E400002-B5A3-F393-E0A9-E50E24DCCA9E".ToLower();

        private CompositeDisposable _compositeDisposable;

        public ReactiveCommand<ManagedScanResult> PeripheralSelectedCommand { get; }

        public ObservableCollection<ManagedScanResult> Peripherals { get; }

        public AboutViewModel()
        {
            _compositeDisposable = new CompositeDisposable();
            Title = "About";
            this.PeripheralSelectedCommand = new ReactiveCommand<ManagedScanResult>().DisposedBy(_compositeDisposable);
            this.PeripheralSelectedCommand.Subscribe(this.OnPeripheralSelected).DisposedBy(_compositeDisposable);
            
            var scanConfig = new ScanConfig
            {
                ServiceUuids = new[] { this.UartService }.ToList()
            };

            var manager = ShinyHost.Resolve<IBleManager>();
            var managedScanner = manager
                .CreateManagedScanner(null, TimeSpan.FromSeconds(5), scanConfig)
                .DisposedBy(_compositeDisposable);
            this.Peripherals = managedScanner.Peripherals;
            managedScanner.Start();
        }

        private void OnPeripheralSelected(ManagedScanResult scanResult)
        {
            IManagedPeripheral managedPeripheral = scanResult.Peripheral.CreateManaged();
            var connectionConfig = new ConnectionConfig { AutoConnect = true };
            managedPeripheral.ConnectWait(connectionConfig).Subscribe(this.OnPeripheralConnected).DisposedBy(_compositeDisposable);
        }

        private void OnPeripheralConnected(ManagedPeripheral managedPeripheral)
        {
            // This method get's called. Also the LED on the Adafruit displays that it is connected.
            managedPeripheral.Read(this.UartService, this.RxdCharacteristic).Subscribe(x => this.OnUartDataReceived(managedPeripheral, x)).DisposedBy(_compositeDisposable);
        }

        private void OnUartDataReceived(ManagedPeripheral managedPeripheral, byte[] data)
        {
            // This never gets called
            string receivedText = Encoding.ASCII.GetString(data);

        }

        public void Dispose()
        {
            _compositeDisposable.Dispose();
        }
    }
```

**Reply by aritchie** [reply]
@E812 please understand that this forums isn't meant for technical support or code review requests.  I do appreciate that you are having troubles, but I already give away the code for free.  And no, you didn't miss anything.  The samples are still something I'm working on.   

Your code looks correct.  I do know managed peripherals work because I'm using them actively in a project right now.  You aren't trapping any errors in your subscribe so I would start there.

**Reply by E812** [reply]
Thanks for your reply. I really appreciate the work you have done and that you respond to questions here. I did not want to complain. I'm not a native English speaker and may not have found the right words. My request was in no way meant to be offensive. If it appeared so to you, then I would like to appologize. 
Your confirmation that I am not completely on the wrong track is already helping me. I will check my error handling and hopefully find the reason why it is not working.

**Reply by aritchie** [reply]
I'm not offended at all.  I can see your github account doesn't have much activity, so I'm making sure you understand the purpose of this forum.  Many users don't bother to consider that OSS is not a product, it is something that most maintainers do in their spare time.  I'm happy if people get value out of the stuff I put out there, but my time isn't free - the source code is.
